client/DESIGN.md

```diff
@@ -0,0 +1,292 @@
+# Treode Client Library Interface
+
+The two major components of the Treode client library interface are the [Cache](#Cache) and the [Transaction](#Transaction).  
+
+Caches store key-value pairs retrieved from the server, as well as the read and value timestamps associated with each pair that Treode provides.  Caches are responsible for maintaining a connection to the DB server.  Users may create multiple caches to work on disjoint sections of the database simultaneously.  
+
+Transactions allows users to locally perform a sequence of database operations and, when finished, optimistically commit the result to the server by writing through the cache.  Users can choose upon creating a Transaction which cache to associate it with.
+
+<a name="Cache"></a>
+# Cache
+
+We provide an overview of the Cache class below, including its primary internal data structure, its public methods, and the form of its HTTP requests and responses.
+
+- Parameters: 
+    1. server: string (required)
+    2. port: string (port=80)
+    3. max\_age: int (max\_age=None)
+    4. no\_cache: bool (no\_cache=False)
+- Public Methods: 
+    1. read(table: string, key: string, read\_timestamp=currentTime, max\_age=None, no\_cache=False): value (JSON)
+    2. write(table: string, op\_list: list): status (string), throws StaleException
+- Internal Objects: 
+    1. HTTP library for establishing/managing connection to server (in Python, urllib3)
+    2. sorted map for maintaining cache (in Python, [skiplist](https://github.com/sepeth/python-skiplist))
+- Exceptions: 
+    1. StaleException
+    2. StaleException(Value_time: int)
+
+## Cache Structure Description
+
+The cache internally maintains (1) a connection to the DB server and (2) a sorted key-value map of cache entries.  
+
+### Database Connection
+
+The cache should utilize an HTTP Request/Response library (e.g., in Python, urllib3) to establish a connection with the DB server and read and write values through its course of use.  We describe in the HTTP Message Formatting section the expected format for DB requests and responses.
+
+[Here are examples](https://urllib3.readthedocs.org/en/latest/) of using the urllib3 library in Python to open and use a connection.
+
+### Sorted Key-Value Map of Cache Entries
+
+The cache is a ("table:key" map) of (sorted value maps), which we explain below.
+
+The keys of the "table:key" map are composed of the concatenation of the table identifier string and the key string; each "table:key" maps to a (sorted value map).  (Note that the cache may store entries from multiple tables.)  
+
+The cache entries themselves are stored in (sorted value maps).  The keys of the (sorted value maps) are derived from the read timestamps of each entry.  Since we want the entries sorted from most-recent to oldest read timestamps, the actual keys are longs with the value 
+
```

Comment:
> keys...are derived from the read timestamps

I think it's easier to derive the keys from the value timestamp. For two reasons. I'll write out examples as `table : key : value-time -> value : read-time` I'll write them using the actual value times in reverse order, rather than writing them out as (Max-time).

#### Finding the value as-of some read-time

Suppose our cache has:

    t : k : 14 -> v2 : 20
    t : k : 6 -> v1 : 10

If our user later asks for `t : k` as-of time 8, we can use `t : k : 8` as our search key. The next entry (ceiling) will be `t : k : 6`. We will get `v1 : 10`. We can see that our cache has no need to fetch from the DB, ever: since 6 (value-time) <= 8 (this read-time) <= 10 (prior read-time), we have an assertion that v1 is the value from time 6 to 10, inclusive.

If our user later asks for `t : k` as-of time 12, we can use `t : k : 12` as our search key. Again, the next entry will be `t : k : 6`, and we again will get `v1 : 10`. Now we can ask, is `12 (this read-time) - 10 (priori read-time) < max-age`? If so, this cache entry is young enough for the cache to provide it as the answer. If not, then the cache needs to request from the DB. When the DB responds, the cache will update the entry for `t : k : 6`, recording the new read-time so that the whole map is now:

    t : k : 14 -> v2 : 20
    t : k : 6 -> v1 : 12

The first map stored the knowledge that `t : k` mapped to v1 from time 6 to 10. The new map stores our expanded knowledge that `t: k` maps to v1 from time 6 to 12.

If we had make the map keys from `t : k : (read time) -> v : (value time)`, our original map would be 

    t : k : 20 -> v2 : 14
    t : k : 10 -> v1 : 6

Now when we want to read as of time 8, we can still see we our okay. We look up the prior (floor) entry for `t : k : 8` and get `t : k : 10`, which maps to `v1 : 6`. We see 8 is between 6 and 10, so we're good.

But when we go to ask for the value as of 12, we have an awkward search procedure. We find the prior entry for `t : k : 12`, and we get `v2 : 14`. Since time 12 is before 14, we know the entry doesn't apply, so we need to check the next entry.

In short, using the value time in the key lets us find the right key as-of this read-time. Using the prior read-time in the key makes it awkward to find the right entry as of this read-time.